### ISSUE TITLE
Weird hook behavior "hookActionProductAdd", need improvements

### ISSUE THREAD

**Issue by iidaloiki** [PrestaShop/PrestaShop#20764]
See @jolelievre amazing summary below https://github.com/PrestaShop/PrestaShop/issues/20764#issuecomment-682521026

<hr>

#### Describe the bug

I'm trying to register hook actionProductAdd. In BO it shows that my module is transplanted in hook, but when I create a product, it does not trigger.

#### Expected behavior

When I create a product, I should get a log that says 'hookActionProductAdd()' which indicates that my hook has been triggered.

#### Steps to Reproduce

1. Create a module
2. Register hook 'actionProductAdd'
3. Add a log into hook:

```
public function hookActionProductAdd($params) {
    Logger::addLog('hookActionProductAdd()', 1, null, null,  null, true);
}
```
4. Create product
5. Check logs - log does not show

#### Additional information

* PrestaShop version: 1.7.6.5
* PHP version: 7.2

**Comment by khouloudbelguith** [comment]
Hi @iibs,

I tried with [ps_featuredproduct](https://github.com/PrestaShop/ps_featuredproducts) module
![image](https://user-images.githubusercontent.com/16067358/91544437-0b896380-e920-11ea-8153-e2b588e3a4db.png)
I tried to edit this [line](https://github.com/PrestaShop/ps_featuredproducts/blob/master/ps_featuredproducts.php#L95)
**From**
`$this->_clearCache('*');`
**To**
`$this->_clearCache('*')`

When I try to save a product, I have this exception
![image](https://user-images.githubusercontent.com/16067358/91544772-7d61ad00-e920-11ea-95ba-b5b15826863a.png)

Thanks to check and feedback.

**Comment by iidaloiki** [comment]
I managed to get same error. I also tried to remove ; from my log, and then create product page crashed even before it rendered. However, I can't still get my log to show.

I also tried renaming my hook to hookAddProduct(), like in ps_featuredproduct, and replanting the module into hook, but it still does not log.

**Comment by khouloudbelguith** [comment]
Ping @PrestaShop/prestashop-core-developers what do you think?

Thanks!

**Comment by jolelievre** [comment]
Hi @iibs 

I tried this hook with the https://github.com/PrestaShop/ps_qualityassurance module, and indeed I don't see the `actionProductAdd` triggered, it does work with `actionProductUpdate` however.

I made some digging into this turns out `actionProductAdd` is not a generic hook used by all ObjectModels, it's a custom one and the only places I was able to find it were when we duplicate a product (both in legacy controller and migrated controller). So maybe the hook is not very well named but anyway that explains why it's not triggered

However you could try one of this two generic hooks
- `actionObjectProductAddBefore`
- `actionObjectProductAddAfter`

I tested both of them and they work as expected.

**However** keep in mind that the product page has a special behaviour, when you click on "New Product" a "ghost" product is automatically and instantaneously created, however it is marked with a temporary state, it is only when you save the product the first time that the product will stop being a ghost/temporary product and will appear in the catalog (wether it's active or not).

So I don't know your use case, but it might not be the product addition hook that you will need, because when products are created from the BackOffice they are always empty shells. Although this might be more useful if it's added by an import, an external module or an API.

## TLDR

The hook is triggered when called, it's just the name is confusing so it's not called when you would expect to, it is only called when a product is duplicated. So I suggest to:
- update the documentation to note this behaviour
- update hook descriptions in the core
- deprecated this hook, replace it with `actionProductDuplicate` and create an alias to the original `actionProductAdd`

If this explanation is satisfactory for you we can either close this issue and open a new one to create the `actionProductDuplicate` or we can update this issue's description and purpose to manage this change.

**Comment by jolelievre** [comment]
By the way I forgot to mention when you duplicate the product (there's a button in the product page), you do see the log :)

**Comment by iidaloiki** [comment]
Hi @jolelievre 

> So I don't know your use case, but it might not be the product addition hook that you will need

This is what I want to achieve:
1. Create product pack
2. Add items into pack, additional info etc.
3. Save product
4. Trigger hook to count product pack quantity based on products included in pack

I have already hooked into `actionProductUpdate`, but that does not trigger when product is saved for the first time. That's why I tried to explore different add hooks.

When I try to transplant `hookActionObjectProductAddAfter`, it's not shown on hook list, even though I have implemented it in code. I would prefer not to reset module, because I have additional tables and columns that include tons of information I don't want to lose.

**Comment by matks** [comment]
@jolelievre has provided an amazing specification of what to be done so I mark this issue as `To Do`

**Comment by jolelievre** [comment]
@iibs I think the hook is not present in the list of transplant because it's a "dynamic hook" so its name was not added in the initial hooks fixtures even though it works, it's not viewable in the database either by default until some module registered to it.

I believe that you can implement a `reset` method in your module When this function doesn't exist then the core interprets a `reset` as `uninstall` + `install` which is what you want to avoid because it would remove your tables

But if you have a custom `reset` method you can make sure all your hooks are correctly registered without removing your data.

Alternatively I think you can also implement `enable` and `disabled` methods which would be responsible to `register` `unregister` your hooks without touching the data in database, this would be another solution for you.

**Comment by boo-code** [comment]
I have written @jolelievre's second item, and I would like a second look at the third before anyone writes that one.

**Item 2, the descriptions.** `actionProductAdd` is dispatched from exactly one place, `ProductDuplicator.php:105`, yet it shipped as:

```xml
<title>Product creation</title>
<description>This hook is displayed after a product is created</description>
```

That is what sends a developer to it, which is how this issue starts. It now describes duplication and names `actionObjectProductAddAfter` for the creation case. Since the titles and descriptions live in the `hook` table and are seeded from the fixture at install, there is an autoupgrade script alongside it for existing shops, conditional on the rows still holding the shipped text so a customised shop keeps its own.

**Item 3, the rename to `actionProductDuplicate` with an alias.** Something the 2020 analysis does not mention: `actionAdminDuplicateAfter` already exists and is dispatched for the same event, two lines below, in the same method:

```php
$this->hookDispatcher->dispatchWithParameters(
    'actionProductAdd',
    ['id_product_old' => $oldProductId, 'id_product' => $newProductId, 'product' => $newProduct]
);

$this->hookDispatcher->dispatchWithParameters(
    'actionAdminDuplicateAfter',
    ['id_product' => $oldProductId, 'id_product_new' => $newProductId]
);
```

So adding `actionProductDuplicate` would leave three names for one dispatch point. The reason nobody uses the one that already says what it means is that it shipped with an empty title and an empty description, which is the other half of what I changed.

Feasibility is not the objection — I checked that a rename would have been safe. `Hook::getHookModuleExecList()` explicitly adds modules registered to a hook's aliases, and `hook_alias.xml` already maps `addproduct` to `actionProductAdd`, so an alias would have kept existing modules working. It is a design question: is a third name wanted, or is describing the two that exist enough?

I have stopped at the descriptions for that reason. If you still want the rename I will write it, alias and upgrade script included; I did not want to add a name to a dispatch point that already has two without asking.

I am at my open pull request limit, so I will link both branches as soon as I have slots.
